applications/OptimizationApplication/custom_utilities/filtering/explicit_vertex_morphing_filter.cpp

```diff
@@ -0,0 +1,311 @@
+#include "utilities/parallel_utilities.h"
+#include "utilities/atomic_utilities.h"
+
+// Application includes
+
+// Include base h
+#include "explicit_vertex_morphing_filter.h"
+
+namespace Kratos {
+
+namespace ExplicitVertexMorphingFilterHelperUtilities
+{
+
+template<class TContainerType>
+const TContainerType& GetContainer(const ModelPart& rModelPart);
+
+template<>
+const ModelPart::NodesContainerType& GetContainer(const ModelPart& rModelPart)
+{
+    return rModelPart.Nodes();
+}
+
+template<>
+const ModelPart::ConditionsContainerType& GetContainer(const ModelPart& rModelPart)
+{
+    return rModelPart.Conditions();
+}
+
+template<>
+const ModelPart::ElementsContainerType& GetContainer(const ModelPart& rModelPart)
+{
+    return rModelPart.Elements();
+}
+
+template <class TEntityType>
+void ComputeWeightForAllNeighbors(
+    double& rSumOfWeights,
+    std::vector<double>& rListOfWeights,
+    const FilterFunction& rFilterFunction,
+    const double VertexMorphingRadius,
+    const EntityPoint<TEntityType>& rDesignPoint,
+    const std::vector<typename EntityPoint<TEntityType>::Pointer>& rNeighbourNodes,
+    const IndexType NumberOfNeighbours)
+{
+    for (IndexType i = 0; i < NumberOfNeighbours; ++i) {
+        const double weight = rFilterFunction.ComputeWeight(rDesignPoint.Coordinates(), rNeighbourNodes[i]->Coordinates(), VertexMorphingRadius);
+        rListOfWeights[i] = weight;
+        rSumOfWeights += weight;
+    }
+}
+
+}; // namespace ExplicitVertexMorphingFilterHelperUtilities
+
+template<class TContainerType>
+ExplicitVertexMorphingFilter<TContainerType>::ExplicitVertexMorphingFilter(
+    const ModelPart& rModelPart,
+    const std::string& rKernelFunctionType,
+    const IndexType MaxNumberOfNeighbours)
+    : mrModelPart(rModelPart),
+      mMaxNumberOfNeighbors(MaxNumberOfNeighbours)
+{
+    mpKernelFunction = Kratos::make_unique<FilterFunction>(rKernelFunctionType);
+    Update();
+}
+
+template<class TContainerType>
+void ExplicitVertexMorphingFilter<TContainerType>::SetFilterRadius(const ContainerExpression<TContainerType>& rContainerExpression)
+{
+    KRATOS_ERROR_IF_NOT(rContainerExpression.GetItemComponentCount() == 1)
+        << "Only scalar values are allowed for the filter radius container expression. "
+        << "Provided contaienr expression = " << rContainerExpression << ".\n";
+
+    KRATOS_ERROR_IF_NOT(&rContainerExpression.GetModelPart() == &mrModelPart)
+        << "Filter radius container expression model part and filter model part mismatch."
+        << "\n\tFilter = " << *this
+        << "\n\tContainerExpression = " << rContainerExpression;
+
+    mpFilterRadiusContainer = rContainerExpression.Clone();
+}
+
+template<class TContainerType>
+void ExplicitVertexMorphingFilter<TContainerType>::ExplicitVertexMorphingFilter::Update()
+{
+    KRATOS_TRY
+
+    BuiltinTimer timer;
+
+    KRATOS_INFO("OptimizationApplication") << "Creating search tree to perform mapping..." << std::endl;
```

Comment:
I would add ExplicitFilter::Update to KRATOS_INFO and ExplicitFilter to all Infos in this class.

Reply:
done :)